services/ui_backend_service/data/cache/custom_flowgraph.py

```diff
@@ -210,25 +202,58 @@ def __str__(self):
         return '\n'.join(str(n) for _, n in sorted((n.func_lineno, n)
                                                    for n in self.nodes.values()))
 
-    def output_dot(self):
-
-        def edge_specs():
-            for node in self.nodes.values():
-                for edge in node.out_funcs:
-                    yield '%s -> %s;' % (node.name, edge)
-
-        def node_specs():
-            for node in self.nodes.values():
-                nodetype = 'join' if node.num_args > 1 else node.type
-                yield '"{0.name}"'\
-                      '[ label = <<b>{0.name}</b> | <font point-size="10">{type}</font>> '\
-                      '  fontname = "Helvetica" '\
-                      '  shape = "record" ];'.format(node, type=nodetype)
-
-        return "digraph {0.name} {{\n"\
-               "rankdir=LR;\n"\
-               "{nodes}\n"\
-               "{edges}\n"\
-               "}}".format(self,
-                           nodes='\n'.join(node_specs()),
-                           edges='\n'.join(edge_specs()))
+    def output_steps(self):
+
+        steps_info = {}
+        graph_structure = []
+
+        def node_to_type(node):
```

Comment:
I'd add a comment explaining why the return value is different from `node.type` and what it used for later

Reply:
> Overall lgtm! just a couple of nitpicks for readability
> 
> Also it seems like we don't have tests for this stuff? How hard would it be to add them (could be a separate PR)

Unfortunately at this moment the cache is not a part of any integration end-to-end testing, there are only unit tests on the cache action level, and in case a cache response is required for other testing, we have usually used patching

Reply:
added a comment to the function.